internal/driver/checkstatuses.go

```diff
@@ -0,0 +1,162 @@
+		return UpWithoutAuth
+	}
+
+	return UpWithAuth
+}
+
+// tcpProbe attempts to make a connection to a specific ip and port list to determine
+// if there is a service listening at that ip+port.
+func (d *Driver) tcpProbe(device sdkModel.Device) bool {
+	proto, ok := device.Protocols[OnvifProtocol]
+	if !ok {
+		d.lc.Warnf("Device %s is missing required %s protocol info, cannot send probe.", device.Name, OnvifProtocol)
+		return false
+	}
+	addr := proto[Address]
+	port := proto[Port]
+
+	if addr == "" || port == "" {
+		d.lc.Warnf("Device %s has no network address, cannot send probe.", device.Name)
+		return false
+	}
+	host := addr + ":" + port
+
+	conn, err := net.DialTimeout("tcp", host, time.Duration(d.config.AppCustom.ProbeTimeoutMillis)*time.Millisecond)
+	if err != nil {
+		d.lc.Debugf("Connection to %s failed when using simple tcp dial, Error: %s ", device.Name, err.Error())
+		return false
+	}
+	defer conn.Close()
+	return true
+}
+
+func (d *Driver) updateDeviceStatus(device sdkModel.Device, status string) error {
+	// todo: maybe have connection levels known as ints, so that way we can log at different levels based on
+	// if the connection level went up or down
+	shouldUpdate := false
+
+	oldStatus := device.Protocols[OnvifProtocol][DeviceStatus]
+	if oldStatus != status {
+		d.lc.Infof("Device status for %s is now %s (used to be %s)", device.Name, status, oldStatus)
+		device.Protocols[OnvifProtocol][DeviceStatus] = status
+		shouldUpdate = true
+	}
+
+	if status != Unreachable {
+		device.Protocols[OnvifProtocol][LastSeen] = time.Now().Format(time.UnixDate)
+		shouldUpdate = true
+	}
+
+	if shouldUpdate {
+		return service.RunningService().UpdateDevice(device)
+	}
+
+	return nil
+}
+
+// taskLoop manages all of our custom background tasks such as checking camera statuses at regular intervals
+func (d *Driver) taskLoop(ctx context.Context) {
+	d.configMu.RLock()
+	interval := d.config.AppCustom.CheckStatusInterval
+	d.configMu.RUnlock()
+	if interval > maxStatusInterval { // check the interval
+		d.lc.Warnf("Status interval of %d seconds is larger than the maximum value of %d seconds. Status interval has been set to the max value.", interval, maxStatusInterval)
+		interval = maxStatusInterval
+	}
+
+	statusTicker := time.NewTicker(time.Duration(interval) * time.Second) // TODO: Support dynamic updates for ticker interval
+
+	defer statusTicker.Stop()
+
+	d.lc.Info("Starting task loop.")
+
+	for {
+		select {
+		case <-ctx.Done():
+			d.lc.Info("Task loop stopped.")
+			return
+		case <-statusTicker.C:
+			start := time.Now()
+			d.checkStatuses() // checks the status of every device
+			d.lc.Debugf("checkStatuses completed in: %v", time.Since(start))
+		}
+	}
+}
+
+// StartTaskLoop runs the taskLoop in the background until cancelled
+func (d *Driver) StartTaskLoop() error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		d.taskLoop(ctx)
+		d.lc.Info("Task loop has exited.")
+	}()
+
+	go func() {
```

Comment:
Can we use the driver's stop func to stop the task loop?
```
func (d *Driver) Stop(force bool) error {
	...
}
```

Reply:
Yes that seems like a good idea.

Reply:
Implemented